gst-libs/gst/androidjni/gstjniamcdirectbuffer.c

```diff
@@ -68,10 +70,25 @@ gst_jni_amc_direct_buffer_render (GstJniAmcDirectBuffer * buf)
 
   if (!buf->released) {
     JNIEnv *env;
-
+    /* From releaseOutputBuffer doc:
+
+       If you render your buffer on a SurfaceView, you can use the timestamp
+       to render the buffer at a specific time (at the VSYNC at or after the
+       buffer timestamp). For this to work, the timestamp needs to be reasonably
+       close to the current System#nanoTime. Currently, this is set as within
+       one (1) second.
+       -----------------------------
+       We are rendering with "now" timestamp because we want the surface to
+       be updated as soon as possible.
+       -----------------------------
+       It's checked that g_get_monotonic_time () * 1000 is the same timestamp as
+       System.nanoTime (), if only with a bit lower precision.
+     */
     env = gst_jni_get_env ();
+
     ret = gst_jni_call_void_method (env, buf->media_codec,
-        buf->release_output_buffer, buf->idx, TRUE);
+        buf->release_output_buffer_ts, buf->idx,
+        g_get_monotonic_time () * 1000);
```

Comment:
It's not the line that is swapping the buffer to be shown, but the line which is telling to put the image from the decoded buffer to decoder's surface as soon as possible. After that in a sink this we'll take a texture from this surface, and copy it to the window's surface with scaling and drawing the black borders.

Reply:
TY, I knew what releaseOutputBuffer in Android is. If we use the overload that takes a timestamp, like we are doing after the change, that is the time at which the amc sink will render to output, as long as the value is within 1s ahead of system time.
So I am very surprised to know that this is working.
Anyway, you stressed that these changes improved the behavior, and I approved as a temporary solution. We should keep investigating for a more correct solution afterwards, as this does not look correct.

Reply:
I am not sure if this has any effect at all, this timestamp is the one used by SurfaceFlinger's buffer queue, so if we are already changing the timestamp in gst_egl_adaptation_swap_buffers this will have no effect as we are already overriding it in the sink.
If we really want to set a timestamp here, this should never be the current system time, but the system time at which we want to buffer to be rendered.

Reply:
In this implementation (egl) we have 2 surfaces: one to which we do swapBuffers and other, to which we do releaseOutputBuffer. We put "now" timestamp to releaseOutputBuffer to say we want the surface to be updated as soon as possible. It's not that surface user will see, but the other one, immediate, from which we are copying the image.